### ISSUE TITLE
cutting and appending FLAC results in wrong FLAC headers

### ISSUE THREAD

**Issue by mbunkus** [mbunkus/mkvtoolnix#953]
**Original reporter**: *[mbunkus](https://github.com/mbunkus)*

There are certain FLAC headers stored in CodecPrivate. These headers include the stream's length, among other things. mkvmerge does not adjust those headers when splitting or appending is involved resulting in wrong length information upon extraction from the Matroska container.

**Comment by mbunkus** [comment]
I have looked into this long and hard. The amount of work that would be required for making this work is insane. I've already spent a lot of hours on refactoring and laying the groundwork, but there's no end in sight.

In addition mkvmerge would need to decode the whole FLAC stream. For this it cannot really use libFLAC because libFLAC's API is a pull-based model which would not fit into mkvmerge's model at all.

I will therefore not fix this bug.

**Comment by mbunkus** [comment]
Some more details that I've just sent to someone via email and fit here nicely as additional details.

The FLAC format is a ridiculously designed. Modifying it requires re-writing most of the headers. Appending is only possible under pretty interesting circumstances (because certain header fields are used when decoding the frames – like the minimum frame size in samples). The header even has seek tables that mkvmerge would have to modify.

On top of all that there's that bloody MD5 checksum of the decoded data. Which means that modifying the tracks without decoding isn't even possible. In order to update the MD5 checksum mkvmerge would have to decode the whole stream. libFLAC's design, though, makes that completely impossible. libFLAC (both the C and C++ versions) have exactly one function for decoding data, and it uses a pull model: libFLAC will call a »read callback« that you provide in order to read more data.

As soon as your read callback returns anything other than success libFLAC thinks that the end of the stream has been reached. If you call that decoding function again then it won't work.

What this means is that libFLAC's design is 100% incompatible with mkvmerge's design.

So basically in order to support appending and splitting FLAC tracks I would have to implement my own FLAC decoder.
